File: war/solutions/hawken.py
# @EXPECTED_RESULTS@: CORRECT

from collections import deque
from sys import stdin

# Read
n = int(stdin.next().strip())
m = int(stdin.next().strip())

p1stack = deque()
p2stack = deque()

p1bet = deque()
p2bet = deque()

p1card = None
p2card = None

if n:
    line = stdin.next()
    [p1stack.append(int(x)) for x in line.strip().split()]
if m:
    line = stdin.next()
    [p2stack.append(int(x)) for x in line.strip().split()]
iter = 0
while len(p1stack) and len(p2stack):
    p1card = p1stack.popleft()
    p2card = p2stack.popleft()
    if p1card == p2card:
        # Put their current cards as bet
        p1bet.append(p1card)
        p2bet.append(p2card)
        # Put three new as bet
        for i in xrange(0,3):
            if not len(p1stack) or not len(p2stack):
                break
            p1bet.append(p1stack.popleft())
            p2bet.append(p2stack.popleft())
    elif p1card > p2card:
        while len(p1bet):
            p1stack.append(p1bet.popleft())

        p1stack.append(p1card)

        while len(p2bet):
            p1stack.append(p2bet.popleft())

        p1stack.append(p2card)
    else:
        while len(p2bet):
            p2stack.append(p2bet.popleft())

        p2stack.append(p2card)

        while len(p1bet):
            p2stack.append(p1bet.popleft())

        p2stack.append(p1card)
    iter = iter + 1
    if iter == 100000:
        break

if (not len(p1stack) and not len (p2stack)) or iter >= 100000:
    print "draw"
elif not len(p1stack):
    print "mike"
else:
    print "you"

File: war/solutions/hawken.java
/*
# @EXPECTED_RESULTS@: CORRECT
*/

import java.util.Queue;
import java.util.LinkedList;
import java.util.Scanner;

public class hawken {
    public static Queue<Integer> qgen(){
        return new LinkedList<Integer>();
    }
    public static void main(String[] args) {
        Scanner in = new Scanner(System.in);

        Queue<Integer> p1stack = qgen(), p2stack = qgen();
        Queue<Integer> p1bet = qgen(), p2bet = qgen();
        int p1card, p2card;

        int n, m;

        n = in.nextInt();
        m = in.nextInt();
        for(int i = 0; i < n; i++)
            p1stack.add(in.nextInt());
        for(int i = 0; i < m; i++)
            p2stack.add(in.nextInt());

        while(p1stack.peek() != null && p2stack.peek() != null) {
            p1card = p1stack.poll();
            p2card = p2stack.poll();
            if(p1card == p2card) {
                p1bet.add(p1card);
                p2bet.add(p2card);
                for(int i = 0; i < 3; i++){
                    if(p1stack.peek() != null)
                        p1bet.add(p1stack.poll());
                    if(p2stack.peek() != null)
                        p2bet.add(p2stack.poll());
                }
            } else if(p1card > p2card) {
                while(p1bet.peek() != null) {
                    p1stack.add(p1bet.poll());
                }
                p1stack.add(p1card);
                while(p2bet.peek() != null) {
                    p1stack.add(p2bet.poll());
                }
                p1stack.add(p2card);
            } else /* if(p1card < p2card) */ {
                while(p2bet.peek() != null) {
                    p2stack.add(p2bet.poll());
                }
                p2stack.add(p2card);
                while(p1bet.peek() != null) {
                    p2stack.add(p1bet.poll());
                }
                p2stack.add(p1card);
            }
        }
        if(p1stack.peek() == null && p2stack.peek() == null) {
            System.out.println("draw");
        } else if(p1stack.peek() != null) {
            System.out.println("you");
        } else /* if(p2stack.peek() != null) */ {
            System.out.println("mike");
        }
    }
}
